### PR TITLE
Eventhistory Service

### PR DESCRIPTION
**Pull request by kobergj**
Backport number 1 from https://github.com/owncloud/ocis/pull/5506

Introduces the eventhistory service

### CODE REVIEW

File: changelog/unreleased/event-history.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Enhancement: Eventhistory service
+
+Introduces the `eventhistory` service. It is a service that is storing events and providing a grpc API to retrieve them
```
Comment (mmattel) [inline changelog/unreleased/event-history.md:3]:
```suggestion
Introduces the `eventhistory` service. It is a service that stores events and provides a grpc API to retrieve them.
```

File: services/eventhistory/README.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+- `mem`: basic inmemory store and the default
+- `ocmem`: advanced inmemory store allowing max size
+- `redis`: stores data in a configured redis cluster
+- `etcd`: stores data in a configured etcd cluster
+- `nats-js`: stores data using nats
+- `noop`: stores nothing. Useful for testing. Not recommended in productive enviroments
+
+## Retrieving
+
+Other services can call the `eventhistory` service via a grpc call to retrieve events. The request must contain the eventid that should be retrieved
```
Comment (mmattel) [inline services/eventhistory/README.md:21]:
```suggestion
# Eventhistory Service

The `eventhistory` consumes all events from the configured event system like NATS, stores them and allows to retrieve them via an eventid.

## Prerequisites

Running the eventhistory service without an event sytem like NATS is not possible.

## Consuming

The `eventhistory` services consumes all events from the configured event sytem.

## Storing

The `eventhistory` stores each consumed event in the configured store. Possible stores are:
- `mem`: basic inmemory store and the default
- `ocmem`: advanced inmemory store allowing max size
- `redis`: stores data in a configured redis cluster
- `etcd`: stores data in a configured etcd cluster
- `nats-js`: stores data using nats
- `noop`: stores nothing. Useful for testing. Not recommended in productive enviroments

## Retrieving

Other services can call the `eventhistory` service via a grpc call to retrieve events. The request must contain the eventid that should be retrieved.
```
The text should include somehow what clears history data. Is it time or retrieval or...?
`Other services can call...` is it only about services or could that include clients?

File: services/eventhistory/pkg/config/config.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+// GRPCConfig defines the available grpc configuration.
+type GRPCConfig struct {
+	Addr      string                 `ocisConfig:"addr" env:"EVENTHISTORY_GRPC_ADDR" desc:"The bind address of the GRPC service."`
+	Namespace string                 `ocisConfig:"-" yaml:"-"`
+	TLS       *shared.GRPCServiceTLS `yaml:"tls"`
+}
+
+// Store configures the store to use
+type Store struct {
+	Type         string        `yaml:"type" env:"EVENTHISTORY_STORE_TYPE" desc:"The type of the store. Supported are: 'mem', 'ocmem', 'etcd', 'redis', 'nats-js', 'noop'. See readme for details"`
```
Comment (mmattel) [inline services/eventhistory/pkg/config/config.go:37]:
```suggestion
	Type         string        `yaml:"type" env:"EVENTHISTORY_STORE_TYPE" desc:"The type of the store. Supported are: 'mem', 'ocmem', 'etcd', 'redis', 'nats-js', 'noop'. See the text description for details."`
```

File: services/eventhistory/pkg/config/config.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	Addresses    string        `yaml:"addresses" env:"EVENTHISTORY_STORE_ADDRESSES" desc:"A comma seperated list of addresses the store should use. No effect on inmemory implementations"`
```
Comment (mmattel) [inline services/eventhistory/pkg/config/config.go:38]:
```suggestion
	Addresses    string        `yaml:"addresses" env:"EVENTHISTORY_STORE_ADDRESSES" desc:"A comma separated list of addresses the store should use. No effect on 'inmemory' implementations."`
```
What is meant with addresses?

File: services/eventhistory/pkg/config/config.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	Database     string        `yaml:"database" env:"EVENTHISTORY_STORE_DATABASE" desc:"The database the store should use"`
```
Comment (mmattel) [inline services/eventhistory/pkg/config/config.go:39]:
```suggestion
	Database     string        `yaml:"database" env:"EVENTHISTORY_STORE_DATABASE" desc:"The database the store should use."`
```
What are the available options?

File: services/eventhistory/pkg/config/config.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	Table        string        `yaml:"table" env:"EVENTHISTORY_STORE_TABLE" desc:"The table the store should use"`
```
Comment (mmattel) [inline services/eventhistory/pkg/config/config.go:40]:
```suggestion
	Table        string        `yaml:"table" env:"EVENTHISTORY_STORE_TABLE" desc:"The table the store should use."`
```
I guess `database table` is meant - or?

File: services/eventhistory/pkg/config/config.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	RecordExpiry time.Duration `yaml:"record_expiry" env:"EVENTHISTORY_RECORD_EXPIRY" desc:"time to life for events in the store"`
```
Comment (mmattel) [inline services/eventhistory/pkg/config/config.go:41]:
```suggestion
	RecordExpiry time.Duration `yaml:"record_expiry" env:"EVENTHISTORY_RECORD_EXPIRY" desc:"Time to life for events in the store."`
```
What is the unit of time?

File: services/eventhistory/pkg/config/config.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	Size         int           `yaml:"size" env:"EVENTHISTORY_STORE_SIZE" desc:"The maximum size for the store. Only applies when using store type 'ocmem'."`
```
Comment (mmattel) [inline services/eventhistory/pkg/config/config.go:42]:
```suggestion
	Size         int           `yaml:"size" env:"EVENTHISTORY_STORE_SIZE" desc:"The maximum size for the store. Only applies when using store type 'ocmem'."`
```
What is the unit of size?

File: services/eventhistory/README.md (hunk 2 above)
Comment (kobergj) [reply services/eventhistory/README.md:21]: Yes true. There is an envvar `EVENTHISTORY_RECORD_EXPIRY` which defines how long events survive in the store. But it should be mentioned here.

File: services/eventhistory/pkg/config/config.go (hunk 5 above)
Comment (kobergj) [reply services/eventhistory/pkg/config/config.go:39]: No available options. It depends on configuration of the database. For `mem` implementation this has no effect. But other implementations might (or might not) allow configuring different databases so keys do not clash. 

Lets say you use an external db like `redis` or `etcd` and have there configured to have a database called `mydb`. If ocis should use `mydb` to store the events it needs to be configured here. 

Important: `Database` and `Table` are optional and only needed if you want data stored in a specific way.

File: services/eventhistory/pkg/config/config.go (hunk 7 above)
Comment (kobergj) [reply services/eventhistory/pkg/config/config.go:41]: As always it is human readable strings like "2s" or "720m"

File: services/eventhistory/pkg/config/config.go (hunk 8 above)
Comment (kobergj) [reply services/eventhistory/pkg/config/config.go:42]: No unit. It is the absolute amount:
```
	// Size configures the maximum capacity of the cache for
	// the "ocmem" implementation, in number of items that the cache can hold per table.
	// You can use 5000 to make the cache hold up to 5000 elements.
	// The parameter only affects to the "ocmem" implementation, the rest will ignore it.
	// If an invalid value is used, the default of 512 will be used instead.
```

File: services/eventhistory/README.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  - `redis`: Stores data in a configured redis cluster.
+  - `etcd`: Stores data in a configured etcd cluster.
+  - `nats-js`: Stores data using nats.
+  - `noop`: Stores nothing. Useful for testing. Not recommended in productive enviroments.
+
+Events stay in the store for 2 weeks by default. Use `EVENTHISTORY_RECORD_EXPIRY` to adjust this value.
+
+## Retrieving
+
+Other services can call the `eventhistory` service via a grpc call to retrieve events. The request must contain the eventid that should be retrieved.
```
Comment (mmattel) [inline services/eventhistory/README.md:27]: > The request must contain the eventid that should be retrieved.

How does one know a eventID?

File: services/eventhistory/pkg/config/config.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+// GRPCConfig defines the available grpc configuration.
+type GRPCConfig struct {
+	Addr      string                 `ocisConfig:"addr" env:"EVENTHISTORY_GRPC_ADDR" desc:"The bind address of the GRPC service."`
+	Namespace string                 `ocisConfig:"-" yaml:"-"`
+	TLS       *shared.GRPCServiceTLS `yaml:"tls"`
+}
+
+// Store configures the store to use
+type Store struct {
+	Type         string        `yaml:"type" env:"EVENTHISTORY_STORE_TYPE" desc:"The type of the store. Supported are: 'mem', 'ocmem', 'etcd', 'redis', 'nats-js', 'noop'. See the text description for details."`
```
Comment (mmattel) [inline services/eventhistory/pkg/config/config.go:37]:
```suggestion
	Type         string        `yaml:"type" env:"EVENTHISTORY_STORE_TYPE" desc:"The type of the eventhistory store. Supported values are: 'mem', 'ocmem', 'etcd', 'redis', 'nats-js', 'noop'. See the text description for details."`
```

File: services/eventhistory/README.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+Running the eventhistory service without an event sytem like NATS is not possible.
+
+## Consuming
+
+The `eventhistory` services consumes all events from the configured event sytem.
+
+## Storing
+
+The `eventhistory` stores each consumed event in the configured store. Possible stores are:
```
Comment (mmattel) [inline services/eventhistory/README.md:15]:
```suggestion
The `eventhistory` service stores each consumed event via the configured store in `EVENTHISTORY_STORE_TYPE`. Possible stores are:
```

File: services/eventhistory/README.md [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+  - `mem`: Basic inmemory store and the default.
+  - `ocmem`: Advanced inmemory store allowing max size.
```
Comment (mmattel) [inline services/eventhistory/README.md:17]:
```suggestion
  - `mem`: Basic in-memory store and the default.
  - `ocmem`: Advanced in-memory store allowing max size.
```

File: services/eventhistory/README.md [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  - `redis`: Stores data in a configured redis cluster.
+  - `etcd`: Stores data in a configured etcd cluster.
+  - `nats-js`: Stores data using nats.
+  - `noop`: Stores nothing. Useful for testing. Not recommended in productive enviroments.
+
+Events stay in the store for 2 weeks by default. Use `EVENTHISTORY_RECORD_EXPIRY` to adjust this value.
```
Comment (mmattel) [inline services/eventhistory/README.md:23]:
```suggestion
1.   Note that in-memory stores are by nature not reboot persistant.
2.   Events stay in the store for 2 weeks by default. Use `EVENTHISTORY_RECORD_EXPIRY` to adjust this value.
```
Added a presistance info and made a numbered list of items.

File: services/eventhistory/README.md [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+The `eventhistory` services consumes all events from the configured event sytem.
+
+## Storing
+
+The `eventhistory` stores each consumed event in the configured store. Possible stores are:
+  - `mem`: Basic inmemory store and the default.
+  - `ocmem`: Advanced inmemory store allowing max size.
+  - `redis`: Stores data in a configured redis cluster.
+  - `etcd`: Stores data in a configured etcd cluster.
+  - `nats-js`: Stores data using nats.
```
Comment (mmattel) [inline services/eventhistory/README.md:20]: Though I heard about redis and etcd, I never heard of `nats-js`. Maybe to add a link for more reading?

File: services/eventhistory/README.md (hunk 9 above)
Comment (kobergj) [reply services/eventhistory/README.md:27]: One cannot. As said it is only other services that can call eventhistory service. Other service will automatically get the id when consuming events.

File: services/eventhistory/README.md (hunk 14 above)
Comment (kobergj) [reply services/eventhistory/README.md]: It's `nats jetstream`. The same technology we are using as eventbus. It also provides key-value store functionality. I added a link.

File: services/eventhistory/pkg/config/config.go [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+type GRPCConfig struct {
+	Addr      string                 `ocisConfig:"addr" env:"EVENTHISTORY_GRPC_ADDR" desc:"The bind address of the GRPC service."`
+	Namespace string                 `ocisConfig:"-" yaml:"-"`
+	TLS       *shared.GRPCServiceTLS `yaml:"tls"`
+}
+
+// Store configures the store to use
+type Store struct {
+	Type         string        `yaml:"type" env:"EVENTHISTORY_STORE_TYPE" desc:"The type of the eventhistory store. Supported values are: 'mem', 'ocmem', 'etcd', 'redis', 'nats-js', 'noop'. See the text description for details."`
+	Addresses    string        `yaml:"addresses" env:"EVENTHISTORY_STORE_ADDRESSES" desc:"A comma separated list of addresses the store should use. No effect on 'inmemory' implementations."`
```
Comment (mmattel) [inline services/eventhistory/pkg/config/config.go:38]:
```suggestion
	Addresses    string        `yaml:"addresses" env:"EVENTHISTORY_STORE_ADDRESSES" desc:"A comma separated list of addresses to access the configured store. This has no effect when 'in-memory' stores are configured."`
```

File: services/eventhistory/README.md [hunk 16]
```diff
@@ -0,0 +1,28 @@
+# Eventhistory Service
+
+The `eventhistory` consumes all events from the configured event system like NATS, stores them and allows to retrieve them via an eventid.
```
Comment (mmattel) [inline services/eventhistory/README.md:3]:
```suggestion
The `eventhistory` consumes all events from the configured event system like NATS, stores them and allows other services to retrieve them via an eventid.
```

File: services/eventhistory/README.md [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+The `eventhistory` services consumes all events from the configured event sytem.
+
+## Storing
+
+The `eventhistory` service stores each consumed event via the configured store in `EVENTHISTORY_STORE_TYPE`. Possible stores are:
+  - `mem`: Basic in-memory store and the default.
+  - `ocmem`: Advanced in-memory store allowing max size.
+  - `redis`: Stores data in a configured redis cluster.
+  - `etcd`: Stores data in a configured etcd cluster.
+  - `nats-js`: Stores data using key-value-store feature of nats jetstream. https://docs.nats.io/nats-concepts/jetstream/key-value-store
```
Comment (mmattel) [inline services/eventhistory/README.md:20]:
```suggestion
  - `nats-js`: Stores data using key-value-store feature of [nats jetstream](https://docs.nats.io/nats-concepts/jetstream/key-value-store)
```

File: services/eventhistory/pkg/config/config.go [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+	Database     string        `yaml:"database" env:"EVENTHISTORY_STORE_DATABASE" desc:"(optional) The database the store should use."`
```
Comment (mmattel) [inline services/eventhistory/pkg/config/config.go:39]:
```suggestion
	Database     string        `yaml:"database" env:"EVENTHISTORY_STORE_DATABASE" desc:"(optional) The database name the configured store should use. This has no effect when 'in-memory' stores are configured."`
```

File: services/eventhistory/pkg/config/config.go [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+	Table        string        `yaml:"table" env:"EVENTHISTORY_STORE_TABLE" desc:"(optional) The table the store should use."`
```
Comment (mmattel) [inline services/eventhistory/pkg/config/config.go:40]:
```suggestion
	Table        string        `yaml:"table" env:"EVENTHISTORY_STORE_TABLE" desc:"(optional) The database table the store should use. This has no effect when 'in-memory' stores are configured."`
```

File: services/eventhistory/pkg/config/config.go [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+	RecordExpiry time.Duration `yaml:"record_expiry" env:"EVENTHISTORY_RECORD_EXPIRY" desc:"Time to life for events in the store. The duration can be set as number followed by a unit identifier like s, m or h. Defaults to '336h' (2 weeks)."`
+	Size         int           `yaml:"size" env:"EVENTHISTORY_STORE_SIZE" desc:"The maximum amount of items in the store. Only applies when using store type 'ocmem'. Defaults to 512."`
```
Comment (mmattel) [inline services/eventhistory/pkg/config/config.go:42]:
```suggestion
	Size         int           `yaml:"size" env:"EVENTHISTORY_STORE_SIZE" desc:"The maximum quantity of items in the store. Only applies when store type 'ocmem' is configured. Defaults to 512."`
```

File: services/eventhistory/README.md [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+  - `noop`: Stores nothing. Useful for testing. Not recommended in productive enviroments.
+
+1.   Note that in-memory stores are by nature not reboot persistant.
+2.   Events stay in the store for 2 weeks by default. Use `EVENTHISTORY_RECORD_EXPIRY` to adjust this value.
```
Comment (mmattel) [inline services/eventhistory/README.md:24]:
```suggestion
2.   Though usually not necessary, a database name and a database table can be configured for event stores that are not of type `in-memory`. These settings are blank by default which means that the standard settings of the configured store applies.
3.   Events stay in the store for 2 weeks by default. Use `EVENTHISTORY_RECORD_EXPIRY` to adjust this value.
```

File: services/eventhistory/README.md [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+
+The `eventhistory` services consumes all events from the configured event sytem.
+
+## Storing
+
+The `eventhistory` service stores each consumed event via the configured store in `EVENTHISTORY_STORE_TYPE`. Possible stores are:
+  - `mem`: Basic in-memory store and the default.
+  - `ocmem`: Advanced in-memory store allowing max size.
+  - `redis`: Stores data in a configured redis cluster.
+  - `etcd`: Stores data in a configured etcd cluster.
```
Comment (mmattel) [inline services/eventhistory/README.md:19]:
```suggestion
  -   `mem`: Basic in-memory store and the default.
  -   `ocmem`: Advanced in-memory store allowing max size.
  -   `redis`: Stores data in a configured redis cluster.
  -   `etcd`: Stores data in a configured etcd cluster.
```
for codacy

File: services/eventhistory/README.md [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+  - `nats-js`: Stores data using key-value-store feature of nats jetstream. https://docs.nats.io/nats-concepts/jetstream/key-value-store
+  - `noop`: Stores nothing. Useful for testing. Not recommended in productive enviroments.
```
Comment (mmattel) [inline services/eventhistory/README.md:21]:
```suggestion
  -   `noop`: Stores nothing. Useful for testing. Not recommended in productive enviroments.
```
for codacy

File: services/eventhistory/README.md [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+
+1.   Note that in-memory stores are by nature not reboot persistant.
```
Comment (mmattel) [inline services/eventhistory/README.md:23]:
```suggestion
1.  Note that in-memory stores are by nature not reboot persistant.
```

File: services/eventhistory/pkg/config/config.go [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+type GRPCConfig struct {
+	Addr      string                 `ocisConfig:"addr" env:"EVENTHISTORY_GRPC_ADDR" desc:"The bind address of the GRPC service."`
+	Namespace string                 `ocisConfig:"-" yaml:"-"`
+	TLS       *shared.GRPCServiceTLS `yaml:"tls"`
+}
+
+// Store configures the store to use
+type Store struct {
+	Type         string        `yaml:"type" env:"EVENTHISTORY_STORE_TYPE" desc:"The type of the eventhistory store. Supported values are: 'mem', 'ocmem', 'etcd', 'redis', 'nats-js', 'noop'. See the text description for details."`
+	Addresses    string        `yaml:"addresses" env:"EVENTHISTORY_STORE_ADDRESSES" desc:"A comma separated list of addresses to access the configured store. This has no effect when 'in-memory' stores are configured."`
```
Comment (mmattel) [inline services/eventhistory/pkg/config/config.go:38]:
```suggestion
	Addresses    string        `yaml:"addresses" env:"EVENTHISTORY_STORE_ADDRESSES" desc:"A comma separated list of addresses to access the configured store. This has no effect when 'in-memory' stores are configured. Note that the behaviour how addresses are used is dependent on the library of the configured store."`
```

**Comment by mmattel** [comment]
Texts and descriptions are ok from my side.

**Review by rhafer — APPROVED** [review]
Looks good to me codewise.

**Pull request merged**